### PR TITLE
feat: Add a 'tutor k8s registry' command

### PR DESCRIPTION
**Pull request by mrtmm**
Add a `tutor k8s registry` command that configures access to a private image registry.
The command creates imagePullSecret(s) for your private registries and adds them to the `default` ServiceAccount in the given namespace.

### CODE REVIEW

**Review by fghaas — COMMENTED** [review]
Looks good at first glance. Please fix the `TODO` in `.github/workflow/tox.yml` and include that modification in this PR. Thanks!

File: tutorregistry/plugin.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    # Create the secret if one does not already exist.
+    try:
+        utils.kubectl(
+            "create", "secret", "docker-registry", "registrykey",
+            f"--docker-server={ docker_registry }",
+            f"--docker-username={ docker_username }",
+            f"--docker-password={ docker_password }",
+            f"--namespace={ namespace }")
+    except exceptions.TutorError:
+        fmt.echo_info("A secret with the same name already exists ...")
```
Comment (mrtmm) [inline tutorregistry/plugin.py:49]: I think here we would also need to have a way to recreate the secret when credentials have changed for example but I am not sure if we should recreate it (delete the old and create a new one) every time by default or add a flag to either force recreation or skip the creation or something else entirely. @fghaas what do you think?

Comment (fghaas) [reply tutorregistry/plugin.py:49]: You probably can judge this better than I. Does recreating the secret every time have any undesirable side effects? If it does not, I am okay with that idea.

**Review by fghaas — COMMENTED** [review]
I have a general question. What this does, currently, is create a single secret for a single Docker registry. However, you can have other Docker image references that _don't_ use the registry set by `DOCKER_REGISTRY`, but specify their own. Arguably, _that_ registry could also require authentication.

Do you want to consider using a dictionary variable named `REGISTRY_CREDENTIALS`, keyed to the registry name instead of the current username and password variables?

As in,

```yaml
REGISTRY_CREDENTIALS:
  'registry.example.com':
    USERNAME: blah
    PASSWORD: foo
  'registry.example.net':
    USERNAME: spam
    PASSWORD: eggs
```

... or something of that nature?

**Comment by mrtmm** [comment]
Hmm, I think having a dictionary for the credentials is a good idea, the `imagePullSecrets` bit in the ServiceAccount is also a list after all and can hold many. In addition, some secrets may require and `email` field as well aside from the username and password. I'll give that a go :)

File: tutorregistry/plugin.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    creds = config["REGISTRY_CREDENTIALS"]
+
+    all_keys = {"imagePullSecrets": []}
+
+    for registry in list(creds.keys()):
+        key_name = creds[registry].get("key_name", "SET_ME")
+        username = creds[registry].get("username", "SET_ME")
+        password = creds[registry].get("password", "SET_ME")
+        # email is optional; depending on the registry
+        email = creds[registry].get("email", "")
```
Comment (fghaas) [inline tutorregistry/plugin.py:48]: To the best of my knowledge, you can iterate through the keys in a dictionary like so:

```python
for registry in creds:
```

But in this case, perhaps you instead want to use:

```suggestion
    for registry, cred in creds.items():
        key_name = cred.get("key_name", "SET_ME")
        username = cred.get("username", "SET_ME")
        password = cred.get("password", "SET_ME")
        # email is optional; depending on the registry
        email = cred.get("email", "")
```

File: tutorregistry/plugin.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            f"--docker-password={ password }",
+            f"--docker-email={ email }" if email else "")
+
+        if 'SET_ME' in create_cmd:
+            raise exceptions.TutorError(
+                "Missing credentials for a private registry. Please "
+                "provide a 'key_name', 'username' and 'password' for each "
+                "of your registries via the 'REGISTRY_CREDENTIALS' definition "
+                "in your 'config.yml' file."
+            )
```
Comment (fghaas) [inline tutorregistry/plugin.py:64]: I may be missing something important here, but on the face of it the way missing credentials are being handled here strikes me as a bit un-Pythonic and clumsy. What do you think of this alternate approach:

```suggestion
    for registry, cred in creds.items():
        try:
            key_name = cred["key_name"]
            username = cred["username"]
            password = cred["password"]
            # email is optional; depending on the registry
            email = cred.get("email", "")
        except KeyError:
            raise exceptions.TutorError(
                "Missing credentials for a private registry. Please "
                "provide a 'key_name', 'username' and 'password' for each "
                "of your registries via the 'REGISTRY_CREDENTIALS' definition "
                "in your 'config.yml' file."
            )
        create_cmd = (
            "create", "secret", "docker-registry", f"{ key_name }",
            f"--namespace={ namespace }",
            f"--docker-server={ registry }",
            f"--docker-username={ username }",
            f"--docker-password={ password }",
            f"--docker-email={ email }" if email else "")
```

(I hope GitHub doesn't mess up the indentation on this one.)

File: README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+Reference documentation can be found [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account)
+
+Configuration
+-------------
+
+* `K8S_NAMESPACE`
+* `DOCKER_REGISTRY`
+* `REGISTRY_USERNAME`
+* `REGISTRY_PASSWORD`
```
Comment (fghaas) [inline README.md:71]: This bit of the `README` still needs an update.

**Review by fghaas — COMMENTED** [review]
This looks good; I have a few minor suggestions. :)

File: tutorregistry/plugin.py (hunk 2 above)
Comment (mrtmm) [reply tutorregistry/plugin.py]: Fixed in the latest change :)

File: tutorregistry/plugin.py (hunk 3 above)
Comment (mrtmm) [reply tutorregistry/plugin.py]: Fixed in the latest change.

File: README.md (hunk 4 above)
Comment (mrtmm) [reply README.md]: Fixed in the latest change.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+* `REGISTRY_CREDENTIALS`
+
+Where the `REGISTRY_CREDENTIALS` configuration value is a dictionary that requires the following structure and values:
+
+    REGISTRY_CREDENTIALS:
+        example1.com:
+            key_name: <key_name>
+            password: <password>
+            username: <username>
+        example2.com:
```
Comment (fghaas) [inline README.md:80]: This is a really tiny nit, but you should not use `example1.com` and `example2.com`. Per RFCs 2607 and 6761, the *only* second-level domain names we are allowed to use as example domain names in documentation are `example.org`, `example.net`, and `example.org`.

Maybe change to:

```suggestion
        registry.example.com:
            key_name: <key_name>
            password: <password>
            username: <username>
        private-registry.example.org:
```

**Review by fghaas — COMMENTED** [review]
I have one more nit and then this is good to go. :)

**Review by fghaas — APPROVED** [review]
This looks good to me now! But I'll hold off on merging for now, per your request.

File: README.md (hunk 5 above)
Comment (mrtmm) [reply README.md]: Done.

File: tutorregistry/plugin.py (hunk 1 above)
Comment (mrtmm) [reply tutorregistry/plugin.py]: Went with recreating secrets every time, I am not aware of any side effects of that at this time.

**Review by fghaas — APPROVED** [review]

**Pull request merged**